File: apps/relay-go/pkg/cryptocards/crypto_test.go
package cryptocards

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"testing"
)

type vectorPlayer struct {
	Seat        int    `json:"seat"`
	PlayerIdHex string `json:"player_id_hex"`
	EntropyHex  string `json:"entropy_hex"`
}

type vectorFile struct {
	Inputs struct {
		GameIdHex               string         `json:"game_id_hex"`
		DeckSize                int            `json:"deck_size"`
		ShuffleAlgorithmVersion int            `json:"shuffle_algorithm_version"`
		Players                 []vectorPlayer `json:"players"`
		RevealPosition          int            `json:"reveal_position"`
	} `json:"inputs"`
}

func loadVector(t *testing.T) vectorFile {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	// pkg/cryptocards -> repo root is 4 hops: pkg/cryptocards ->
	// pkg -> relay-go -> apps -> root.
	path := filepath.Join(wd, "..", "..", "..", "..", "spec", "test-vectors", "mental-poker.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read vector %q: %v", path, err)
	}
	var v vectorFile
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("parse vector: %v", err)
	}
	return v
}

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("hex.DecodeString(%q): %v", s, err)
	}
	return b
}

func TestCommitEntropyRoundtrip(t *testing.T) {
	v := loadVector(t)
	p := v.Inputs.Players[0]
	entropy := mustHex(t, p.EntropyHex)
	playerId := mustHex(t, p.PlayerIdHex)
	gameId := mustHex(t, v.Inputs.GameIdHex)

	first, err := CommitEntropy(entropy, playerId, gameId)
	if err != nil {
		t.Fatalf("CommitEntropy: %v", err)
	}
	second, err := CommitEntropy(entropy, playerId, gameId)
	if err != nil {
		t.Fatalf("CommitEntropy(again): %v", err)
	}
	if hex.EncodeToString(first) != hex.EncodeToString(second) {
		t.Errorf("CommitEntropy is non-deterministic")
	}
	if len(first) != 32 {
		t.Errorf("commitment length: want 32, got %d", len(first))
	}

	ok, err := VerifyEntropyReveal(first, entropy, playerId, gameId)
	if err != nil {
		t.Fatalf("VerifyEntropyReveal: %v", err)
	}
	if !ok {
		t.Error("VerifyEntropyReveal: want true, got false")
	}

	bad := append([]byte{}, entropy...)
	bad[0] ^= 0xff
	ok, err = VerifyEntropyReveal(first, bad, playerId, gameId)
	if err != nil {
		t.Fatalf("VerifyEntropyReveal(tampered): %v", err)
	}
	if ok {
		t.Error("VerifyEntropyReveal(tampered): want false, got true")
	}
}

func TestCombineEntropyDeterministicOrderSensitive(t *testing.T) {
	v := loadVector(t)
	entropies := make([][]byte, len(v.Inputs.Players))
	for i, p := range v.Inputs.Players {
		entropies[i] = mustHex(t, p.EntropyHex)
	}
	a, err := CombineEntropy(entropies)
	if err != nil {
		t.Fatalf("CombineEntropy: %v", err)
	}
	b, err := CombineEntropy(entropies)
	if err != nil {
		t.Fatalf("CombineEntropy(again): %v", err)
	}
	if hex.EncodeToString(a) != hex.EncodeToString(b) {
		t.Error("CombineEntropy is non-deterministic")
	}
	swapped := [][]byte{entropies[1], entropies[0], entropies[2]}
	c, err := CombineEntropy(swapped)
	if err != nil {
		t.Fatalf("CombineEntropy(swapped): %v", err)
	}
	if hex.EncodeToString(c) == hex.EncodeToString(a) {
		t.Error("CombineEntropy order-insensitive")
	}
}

func TestShuffleIsPermutationDeterministic(t *testing.T) {
	v := loadVector(t)
	entropies := make([][]byte, len(v.Inputs.Players))
	for i, p := range v.Inputs.Players {
		entropies[i] = mustHex(t, p.EntropyHex)
	}
	combined, err := CombineEntropy(entropies)
	if err != nil {
		t.Fatalf("CombineEntropy: %v", err)
	}
	a, err := ShuffleCanonicalDeck(combined, v.Inputs.DeckSize)
	if err != nil {
		t.Fatalf("Shuffle: %v", err)
	}
	b, err := ShuffleCanonicalDeck(combined, v.Inputs.DeckSize)
	if err != nil {
		t.Fatalf("Shuffle(again): %v", err)
	}
	if len(a) != len(b) {
		t.Fatalf("shuffle length differs: %d vs %d", len(a), len(b))
	}
	for i := range a {
		if a[i] != b[i] {
			t.Errorf("shuffle non-deterministic at %d: %d vs %d", i, a[i], b[i])
		}
	}
	sorted := append([]int{}, a...)
	sort.Ints(sorted)
	for i, x := range sorted {
		if x != i {
			t.Errorf("shuffle not a permutation: sorted[%d]=%d", i, x)
		}
	}
}

func TestDeckCommitmentAndRevealProofRoundtrip(t *testing.T) {
	v := loadVector(t)
	entropies := make([][]byte, len(v.Inputs.Players))
	for i, p := range v.Inputs.Players {
		entropies[i] = mustHex(t, p.EntropyHex)
	}
	combined, err := CombineEntropy(entropies)
	if err != nil {
		t.Fatalf("CombineEntropy: %v", err)
	}
	dc, err := BuildDeckCommitment(combined, v.Inputs.DeckSize, v.Inputs.ShuffleAlgorithmVersion)
	if err != nil {
		t.Fatalf("BuildDeckCommitment: %v", err)
	}
	honest := dc.PerPosition[v.Inputs.RevealPosition]
	ok, err := VerifyRevealProof(honest.Commitment, RevealProof{
		Position:  honest.Position,
		Ordinal:   honest.Ordinal,
		CardNonce: honest.CardNonce,
		DeckNonce: honest.DeckNonce,
	})
	if err != nil {
		t.Fatalf("VerifyRevealProof: %v", err)
	}
	if !ok {
		t.Error("VerifyRevealProof(honest): want true, got false")
	}
	ok, err = VerifyRevealProof(honest.Commitment, RevealProof{
		Position:  honest.Position,
		Ordinal:   (honest.Ordinal + 1) % 52,
		CardNonce: honest.CardNonce,
		DeckNonce: honest.DeckNonce,
	})
	if err != nil {
		t.Fatalf("VerifyRevealProof(forged): %v", err)
	}
	if ok {
		t.Error("VerifyRevealProof(forged): want false, got true")
	}
}

func TestDeckCommitmentHashBindsEntropy(t *testing.T) {
	v := loadVector(t)
	entropies := make([][]byte, len(v.Inputs.Players))
	for i, p := range v.Inputs.Players {
		entropies[i] = mustHex(t, p.EntropyHex)
	}
	combinedA, _ := CombineEntropy(entropies)
	dcA, _ := BuildDeckCommitment(combinedA, 52, 1)

	flipped := append([]byte{}, entropies[0]...)
	flipped[0] ^= 0xff
	combinedB, _ := CombineEntropy([][]byte{flipped, entropies[1], entropies[2]})
	dcB, _ := BuildDeckCommitment(combinedB, 52, 1)

	if hex.EncodeToString(dcA.DeckCommitmentHash) == hex.EncodeToString(dcB.DeckCommitmentHash) {
		t.Error("deck commitment hash insensitive to entropy")
	}
}

File: packages/crypto-cards/__tests__/crypto.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';
import {
  buildDeckCommitment,
  combineEntropy,
  commitEntropy,
  fromHex,
  shuffleCanonicalDeck,
  toHex,
  verifyEntropyReveal,
  verifyRevealProof,
} from '../src/index.js';

interface VectorPlayer {
  readonly seat: number;
  readonly player_id_hex: string;
  readonly entropy_hex: string;
}
interface VectorFile {
  readonly inputs: {
    readonly game_id_hex: string;
    readonly deck_size: number;
    readonly shuffle_algorithm_version: number;
    readonly players: readonly VectorPlayer[];
    readonly reveal_position: number;
  };
}

const vectorPath = resolve(
  dirname(fileURLToPath(import.meta.url)),
  '../../../spec/test-vectors/mental-poker.json',
);
const vector = JSON.parse(readFileSync(vectorPath, 'utf8')) as VectorFile;

describe('mental-poker crypto — commit/reveal/verify roundtrip', () => {
  it('commitEntropy is deterministic and verifyEntropyReveal accepts the matching plaintext', async () => {
    const p = vector.inputs.players[0]!;
    const entropy = fromHex(p.entropy_hex);
    const playerId = fromHex(p.player_id_hex);
    const gameId = fromHex(vector.inputs.game_id_hex);

    const first = await commitEntropy(entropy, playerId, gameId);
    const second = await commitEntropy(entropy, playerId, gameId);
    expect(toHex(first)).toEqual(toHex(second));
    expect(first.byteLength).toBe(32);

    expect(await verifyEntropyReveal(first, entropy, playerId, gameId)).toBe(true);

    // Tampered entropy must reject.
    const bad = new Uint8Array(entropy);
    bad[0] ^= 0xff;
    expect(await verifyEntropyReveal(first, bad, playerId, gameId)).toBe(false);
  });

  it('combineEntropy is deterministic and order-sensitive', async () => {
    const entropies = vector.inputs.players.map((p) => fromHex(p.entropy_hex));
    const a = await combineEntropy(entropies);
    const b = await combineEntropy(entropies);
    expect(toHex(a)).toEqual(toHex(b));
    expect(a.byteLength).toBe(32);

    // Swapping the order MUST change the result.
    const swapped = [entropies[1]!, entropies[0]!, entropies[2]!];
    const c = await combineEntropy(swapped);
    expect(toHex(c)).not.toEqual(toHex(a));
  });

  it('shuffleCanonicalDeck yields a permutation of 0..deckSize-1', async () => {
    const entropies = vector.inputs.players.map((p) => fromHex(p.entropy_hex));
    const combined = await combineEntropy(entropies);
    const deck = await shuffleCanonicalDeck(combined, vector.inputs.deck_size);
    expect(deck).toHaveLength(vector.inputs.deck_size);
    const sorted = [...deck].sort((a, b) => a - b);
    for (let i = 0; i < sorted.length; i++) {
      expect(sorted[i]).toBe(i);
    }
  });

  it('shuffleCanonicalDeck is deterministic for the same combined entropy', async () => {
    const entropies = vector.inputs.players.map((p) => fromHex(p.entropy_hex));
    const combined = await combineEntropy(entropies);
    const a = await shuffleCanonicalDeck(combined, 52);
    const b = await shuffleCanonicalDeck(combined, 52);
    expect(a).toEqual(b);
  });

  it('buildDeckCommitment + verifyRevealProof: accepts an honest open, rejects a forged ordinal', async () => {
    const entropies = vector.inputs.players.map((p) => fromHex(p.entropy_hex));
    const combined = await combineEntropy(entropies);
    const dc = await buildDeckCommitment(combined, vector.inputs.deck_size, vector.inputs.shuffle_algorithm_version);

    const pos = vector.inputs.reveal_position;
    const honest = dc.perPosition[pos]!;
    const ok = await verifyRevealProof(honest.commitment, {
      position: honest.position,
      ordinal: honest.ordinal,
      cardNonce: honest.cardNonce,
      deckNonce: honest.deckNonce,
    });
    expect(ok).toBe(true);

    // Forge the ordinal — verifier must reject.
    const forged = await verifyRevealProof(honest.commitment, {
      position: honest.position,
      ordinal: (honest.ordinal + 1) % 52,
      cardNonce: honest.cardNonce,
      deckNonce: honest.deckNonce,
    });
    expect(forged).toBe(false);
  });

  it('deckCommitmentHash binds the entire shuffled deck (changing one entropy changes the hash)', async () => {
    const entropies = vector.inputs.players.map((p) => fromHex(p.entropy_hex));
    const combinedA = await combineEntropy(entropies);
    const dcA = await buildDeckCommitment(combinedA, 52, 1);

    const flipped = new Uint8Array(entropies[0]!);
    flipped[0] ^= 0xff;
    const combinedB = await combineEntropy([flipped, entropies[1]!, entropies[2]!]);
    const dcB = await buildDeckCommitment(combinedB, 52, 1);

    expect(toHex(dcA.deckCommitmentHash)).not.toEqual(toHex(dcB.deckCommitmentHash));
  });
});
